internal/getproviders/package_authentication.go

```diff
@@ -65,29 +92,192 @@ func NewArchiveChecksumAuthentication(wantSHA256Sum [sha256.Size]byte) PackageAu
+		WantSHA256Sum: wantSHA256Sum,
+	}
+}
+
+func (m matchingChecksumAuthentication) AuthenticatePackage(meta PackageMeta, location PackageLocation) (*PackageAuthenticationResult, error) {
+	if _, ok := meta.Location.(PackageHTTPURL); !ok {
+		// A source should not use this authentication type for non-HTTP
+		// source locations.
+		return nil, fmt.Errorf("cannot verify matching checksum for non-HTTP location %s", meta.Location)
+	}
+
+	// Find the checksum in the list with matching filename. The document is
+	// in the form "0123456789abcdef filename.zip".
+	filename := []byte(m.Filename)
+	var checksum []byte
+	for _, line := range bytes.Split(m.Document, []byte("\n")) {
+		parts := bytes.Fields(line)
+		if len(parts) > 1 && bytes.Equal(parts[1], filename) {
+			checksum = parts[0]
+			break
+		}
+	}
+	if checksum == nil {
+		return nil, fmt.Errorf("checksum list has no SHA-256 hash for %q", m.Filename)
+	}
+
+	// Decode the ASCII checksum into a byte array for comparison.
+	var gotSHA256Sum [sha256.Size]byte
+	if _, err := hex.Decode(gotSHA256Sum[:], checksum); err != nil {
+		return nil, fmt.Errorf("checksum list has invalid SHA256 hash %q: %s", string(checksum), err)
+	}
+
+	// If thee checksums don't match, authentication fails.
+	if !bytes.Equal(gotSHA256Sum[:], m.WantSHA256Sum[:]) {
+		return nil, fmt.Errorf("checksum list has unexpected SHA-256 hash %x (expected %x)", gotSHA256Sum, m.WantSHA256Sum[:])
+	}
+
+	// Success! But this doesn't result in any real authentication, only a
+	// lack of authentication errors, so we return a nil result.
+	return nil, nil
+}
+
+type signatureAuthentication struct {
+	Document  []byte
+	Signature []byte
+	Keys      []SigningKey
+}
+
+// NewSignatureAuthentication returns a PackageAuthentication implementation
+// that verifies the cryptographic signature for a package against a given key.
+func NewSignatureAuthentication(document, signature []byte, keys []SigningKey) PackageAuthentication {
+	return signatureAuthentication{
+		Document:  document,
+		Signature: signature,
+		Keys:      keys,
+	}
+}
+
+func (s signatureAuthentication) AuthenticatePackage(meta PackageMeta, location PackageLocation) (*PackageAuthenticationResult, error) {
+	if _, ok := location.(PackageLocalArchive); !ok {
+		// A source should not use this authentication type for non-archive
+		// locations.
+		return nil, fmt.Errorf("cannot check archive hash for non-archive location %s", location)
 	}
-	return nil
+
+	if _, ok := meta.Location.(PackageHTTPURL); !ok {
+		// A source should not use this authentication type for non-HTTP source
+		// locations.
+		return nil, fmt.Errorf("cannot check archive hash for non-HTTP location %s", meta.Location)
+	}
+
+	// Attempt to verify the signature using each of the keys returned by the
+	// registry. Note: currently the registry only returns one key, but this
+	// may change in the future. We must check each key in turn to find the
+	// matching signing entity before proceeding.
```

Comment:
Instead of iterating over the keys and checking each, you could also iterate over the keys and build an `openpgp.EntityList`, and then use that as the keyring. Pseudo-code:

```go
var keyring openpgp.EntityList
for _, key := range s.Keys {
    entities, err := openpgp.ReadArmoredKeyRing(strings.NewReader(key.ASCIIArmor))

    for _, entity := range entities {
        keyring = append(keyring, entity)
    }
}

signer, err := openpgp.CheckDetachedSignature(keyring, bytes.NewReader(s.Document), bytes.NewReader(s.Signature))
```

I think the way you're doing it is fine too, just wanted to point this out as a possible solution given the behavior of the openpgp library.

Reply:
The reason I didn't take this approach is because I couldn't figure out how to then check the trust signature for the signing key given only an `openpgp.Entity`. Do you know of a way to do that (without just creating a map of `Entity` to ASCII armor)?

Reply:
Good point! Probably not worth keeping a mapping vs the current implementation.